Source/DafnyLanguageServer/Workspace/TextDocumentLoader.cs

```diff
@@ -221,57 +209,60 @@ bool loadCanceled
-        })).Replay();
-      result.Connect();
+      foreach (var implementationTask in implementationTasks) {
+        Verify(entry, document, implementationTask, cancellationToken);
+      }
+
+      if (!implementationTasks.Any()) {
+        entry.EndVerification();
+      }
+
+      var lastDocument = await entry.LastDocument;
 
       if (VerifierOptions.GutterStatus) {
-        ReportRealtimeDiagnostics(document, result, cancellationToken);
+        document.GutterProgressReporter!.ReportRealtimeDiagnostics(true, lastDocument);
       }
 
-      var _ = NotifyStatusAsync(document.TextDocumentItem, result.DefaultIfEmpty(document), cancellationToken);
-      return result;
+      NotifyStatusAsync(document.TextDocumentItem, lastDocument, cancellationToken);
     }
 
-    public IObservable<DafnyDocument> Verify(DafnyDocument dafnyDocument, IImplementationTask implementationTask,
+    public bool Verify(IDocumentEntry entry, DafnyDocument dafnyDocument, IImplementationTask implementationTask,
       CancellationToken cancellationToken) {
 
       var statusUpdates = implementationTask.TryRun();
       if (statusUpdates == null) {
+
         if (VerifierOptions.GutterStatus && implementationTask.CacheStatus is Completed completedCache) {
           foreach (var result in completedCache.Result.VCResults) {
-            dafnyDocument.GutterProgressReporter!.ReportAssertionBatchResult(new AssertionBatchResult(implementationTask.Implementation, result));
+            dafnyDocument.GutterProgressReporter!.ReportAssertionBatchResult(
+              new AssertionBatchResult(implementationTask.Implementation, result));
           }
-          dafnyDocument.GutterProgressReporter!.ReportEndVerifyImplementation(implementationTask.Implementation, completedCache.Result);
-        }
-        return Observable.Empty<DafnyDocument>();
-      }
 
-      cancellationToken.Register(implementationTask.Cancel);
-      return GetVerifiedDafnyDocuments(dafnyDocument, implementationTask, statusUpdates);
-    }
+          dafnyDocument.GutterProgressReporter!.ReportEndVerifyImplementation(implementationTask.Implementation,
+            completedCache.Result);
+        }
 
-    private IObservable<DafnyDocument> GetVerifiedDafnyDocuments(DafnyDocument document, IImplementationTask implementationTask,
-      IObservable<IVerificationStatus> statusUpdates) {
+        return false;
+      }
 
-      var result = statusUpdates.Select(boogieStatus => HandleStatusUpdate(document, implementationTask, boogieStatus)).Replay();
-      result.Connect(); // Immediately start processing the status updates.
+      entry.RestartVerification();
+      statusUpdates.ObserveOn(dafnyDocument.UpdateScheduler).Subscribe(
+        update => HandleStatusUpdate(entry, dafnyDocument, implementationTask, update),
+        () => {
+          if (dafnyDocument.VerificationTasks!.All(t => t.IsIdle)) {
+            entry.EndVerification();
+          }
```

Comment:
Is it safe to iterate on the tasks without locking here?

Reply:
The tasks are only set once during the translation phase. The `.IsIdle` value of the tasks can change after that, but once the final running task reaches this OnComplete handler, no `.IsIdle` changes are still occurring.